design/StableRegions.md

```diff
@@ -0,0 +1,298 @@
+
+For disaster recovery, we can **also** rebuild all of the region objects from data in stable memory.
+
+We use the `block-region` and `region` tables in stable memory to rebuild the regions' objects:
+
+ - The `region` table gives a size for each region, saying how large each vector for each region needs to be.
+ - The `block-region` table gives a relative position and region ID for each block ID.
+
+Once each regions' vectors have been allocated, the block-region table says how to fill them, one entry at a time.
+Unlike the Rust design, vector entries can be populated out-of-order.
+
+
+### Special (reserved) regions
+
+  - Region 0 -- Anonymous region, for supporting the legacy API that we have today, which lacks `Region` values.
+  - Region 1 -- "Reclaimed blocks" region that consists of blocks reclaimed from GC'd regions.
+  - Regions 2-15 -- Future use by Motoko RTS (TBD).
+
+### Overview of GC support (future work)
+
+- Regions are reprensented (see special subsection) with heap objects that are `stable`, but mutable.
+- They have special GC headers to recognize their special structure.
+- The Region table keeps track of which available IDs are in use as Region heap values are GC'd.
+
+#### Region allocation, re-allocation
+
+A region with a particular ID is first initially allocated by a global, bumped counter reaching it.
+
+When that happens, the region table is marked (TO DO) indicating that region is in use.
+
+The region heap object with that ID witnesses its lifetime, and is reclaimed by GC when it goes out of scope.
+
+When the bump counter reaches its maximum value, the allocation algorithm changes to scanning the regions,
+looking for the first available one (with minimum ID).
+
+#### Region reclamation
+
+A heap object with a certain region ID lives until GC reclaims it, where the region table is re-marked (TO DO) to indicate that the region is free again.
+
+Before reusing that region, each of its block are added to the special "free blocks" region.
+
+
+### Migration from earlier designs into region system
+
+#### Stable memory format versions
+
+The first 32 bits of stable memory record a version number.
+
+Including this design, there are three possible verions (`0`, `1`, or `2`):
+
+ 0. Stable vars only.
+ 1. Stable vars *plus* direct access to IC0 API, including `grow`.
+ 2. Region system, where direct access still works through region zero.
+
+In the first cases (`0`) and (`1`), we wish to migrate into the region system (`2`), with its own internal versioning.
+
+To do so, we will read the existing version during upgrades, and apply installation logic that we call "migration".
+
+##### Version 0 migration.
+
+To migrate from version 0 to version 2, there is nothing additional to do for existing data.
+
+The region system detects this case by measuring the zero-sized stable memory during its initialization.
+
+##### Version 1 migration: Version 1 stable memory *as* region 0 in Version 2.
```

Comment:
@crusso Will focus on authoring this logic in the RTS.

I see where to trigger it in `destabilize`, but writing the exact Wasm combinators for the new case of the condition may take some practice for me.

Reply:
The pseudocode functions in design/StableMemory.md should match what the compiler is emitting (see comments in compile.ml). Might be worth understanding/adapting the pseudocode first before attempting the real deal.

Reply:
Maybe it would be easier to write the special migration function in Rust, up to you. I f you need to copy/move some pages, maybe just heap  allocate a blob and write into/copy out from that. The GC will free it later.

Reply:
> Maybe it would be easier to write the special migration function in Rust, up to you.

Yep, that's been my plan.  I've been trying to figure out where to trigger it (the condition we are discussing about version numbers and sizes, etc).

Reply:
Good to know about the heap being available to alloc (with GC freeing things still)